hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryStrategyFactory.java

```diff
@@ -73,4 +75,44 @@
      * @return a set of expected configuration properties
      */
     Collection<PropertyDefinition> getConfigurationProperties();
+
+    /**
+     * Checks whether the given discovery strategy may be applied with no additional config to the environment in which Hazelcast
+     * is currently running.
+     * <p>
+     * Used by the auto detection mechanism to decide which strategy should be used.
+     */
+    default boolean isAutoDetectionApplicable() {
+        return false;
+    }
+
+    /**
+     * Level of the discovery strategy.
+     */
+    default DiscoveryStrategyLevel discoveryStrategyLevel() {
+        return UNKNOWN;
+    }
+
+    /**
+     * Level of the discovery strategy.
+     * <p>
+     * Discovery strategies can have different levels. They can be at the level of Cloud Virtual Machines, for example, AWS EC2
+     * Instance or GCP Virtual Machine. They can also be at the level of some specific platform or framework, like Kubernetes.
+     * <p>
+     * It decides on the priority in the auto detection mechanism, because you can have a Kubernetes environment installed on
+     * AWS EC2 Instances and then Hazelcast Kubernetes Discovery Strategy should take precedence over AWS Discovery Strategy.
+     */
+    enum DiscoveryStrategyLevel {
```

Comment:
This part deserves a more detailed JavaDoc. It's a public API after one. As a SPI implementer this is the first question I have to ask:  What is the system going to prefer? I the strategy with the higher ordinals will be picked, but it should be explicit. Also I am not quite sure what the 'PLATFORM' level means.

Reply:
Added. PTAL